src/Prefetch.cpp

```diff
@@ -0,0 +1,348 @@
+        } else {   // Use element size for inputs that don't have static types
+            Expr elem_size_var = Variable::make(Int(32), elem_size_name);
+            elem_size_bytes = elem_size_var;
+        }
+
+        std::ostringstream ss; ss << t;
+        string type_name = ss.str();
+        debug(dbg_prefetch1) << "  prefetch #" << ptmp << ": "
+                             << varname << " ("
+                             << (has_static_type ? type_name : elem_size_name)
+                             << ", dims:" << dims << ")\n";
+
+        for (int i = 0; i < dims; i++) {
+            debug(dbg_prefetch3) << "    ---\n";
+            debug(dbg_prefetch3) << "    box[" << i << "].min: " << box[i].min << "\n";
+            debug(dbg_prefetch3) << "    box[" << i << "].max: " << box[i].max << "\n";
+        }
+        debug(dbg_prefetch3) << "    ---------\n";
+
+        // TODO: Opt: check box if it should be prefetched?
+        // TODO       - Only prefetch if varying by ivar_name?
+        // TODO       - Don't prefetch if "small" all constant dimensions?
+        // TODO         e.g. see: camera_pipe.cpp corrected matrix(4,3)
+
+        string pstr = std::to_string(ptmp++);
+        string varname_prefetch_buf = varname + "_prefetch_" + pstr + "_buf";
+        Expr var_prefetch_buf = Variable::make(Int(32), varname_prefetch_buf);
+
+        // Establish the variables for buffer strides, box min & max
+        vector<Expr> stride_var(dims);
+        vector<Expr> extent_var(dims);
+        vector<Expr> min_var(dims);
+        vector<Expr> max_var(dims);
+        for (int i = 0; i < dims; i++) {
+            string istr = std::to_string(i);
+            // string extent_name = varname + ".extent." + istr;
+            string stride_name = varname + ".stride." + istr;
+            string extent_name = varname + ".extent." + istr;
+            string min_name = varname + "_prefetch_" + pstr + "_min_" + istr;
+            string max_name = varname + "_prefetch_" + pstr + "_max_" + istr;
+
+            stride_var[i] = Variable::make(Int(32), stride_name);
+            extent_var[i] = Variable::make(Int(32), extent_name);
+            min_var[i] = Variable::make(Int(32), min_name);
+            max_var[i] = Variable::make(Int(32), max_name);
+
+            // Record let assignments
+            // except for stride, already defined elsewhere
+            plets.push_back(make_pair(min_name, box[i].min));
+            plets.push_back(make_pair(max_name, box[i].max));
+        }
+
+        // This box should not be prefetched
+        if (!do_prefetch) {
+            debug(dbg_prefetch1) << "  Info: not prefetching " << varname << "\n";
+            return;
+        }
+
+        // Create a buffer_t object for this prefetch.
+        vector<Expr> args(dims*3 + 2);
+
+        Expr first_elem = Load::make(t, varname, 0, BufferPtr(), Parameter());
+        args[0] = Call::make(Handle(), Call::address_of, {first_elem}, Call::PureIntrinsic);
```

Comment:
This looks like it would be simpler to construct the Expr vector with push_back.